File: examples/spec/example_spec.py
from dataclasses import dataclass
import abc
import typing


# SimpleStruct
@dataclass
class SimpleStruct:
    bool_value: bool
    i8_value: int
    i64_value: typing.Optional[int] = None
    string_value: typing.Optional[str] = None
    bytes_value: typing.Optional[bytes] = None
    string_to_string: typing.Optional[typing.Dict[str, str]] = None
    key_values: typing.Optional["KeyValue"] = None
    children_container: typing.Optional["Container"] = None
    children: typing.Optional[typing.List["SimpleStruct"]] = None

    def to_dict(self):
        result = {}
    
        # bool_value
        result["bool_value"] = self.bool_value
    
        # i8_value
        result["i8_value"] = self.i8_value
    
        # i64_value
        result["i64_value"] = self.i64_value
    
        # string_value
        result["string_value"] = self.string_value
    
        # bytes_value
        result["bytes_value"] = self.bytes_value
    
        # string_to_string
        if self.string_to_string is None:
            result["string_to_string"] = None
        else:
            string_to_string_tmp = {}
            for key, item in self.string_to_string.items():
                item_tmp = item
                string_to_string_tmp[key] = item_tmp
            
            result["string_to_string"] = string_to_string_tmp
    
        # key_values
        if self.key_values is None:
            result["key_values"] = None
        else:
            key_values_tmp = {}
            for key, item in self.key_values.items():
                item_tmp = list(item)
                key_values_tmp[key] = item_tmp
            
            result["key_values"] = key_values_tmp
    
        # children_container
        if self.children_container is None:
            result["children_container"] = None
        else:
            children_container_tmp = []
            for item in self.children_container:
                item_tmp = item.to_dict()
                children_container_tmp.append(item_tmp)
            
            result["children_container"] = children_container_tmp
    
        # children
        if self.children is None:
            result["children"] = None
        else:
            children_tmp = []
            for item in self.children:
                item_tmp = item.to_dict()
                children_tmp.append(item_tmp)
            
            result["children"] = children_tmp
        return result
    

    @staticmethod
    def from_dict(d):
        
        # bool_value
        bool_value = d["bool_value"]
        
        # i8_value
        i8_value = d["i8_value"]
        
        # i64_value
        i64_value = d.get("i64_value", None)
        
        # string_value
        string_value = d.get("string_value", None)
        
        # bytes_value
        bytes_value = None
        if item := d.get("bytes_value"):
            bytes_value = bytes(item)
        
        # string_to_string
        string_to_string = None
        if item := d.get("string_to_string"):
            string_to_string = {}
            for key, item in item.items():
                item_tmp = item
                string_to_string[key] = item_tmp
            
        
        # key_values
        key_values = None
        if item := d.get("key_values"):
            key_values = {}
            for key, item in item.items():
                item_tmp = bytes(item)
                key_values[key] = item_tmp
            
        
        # children_container
        children_container = None
        if item := d.get("children_container"):
            children_container = []
            for item in item:
                item_tmp = SimpleStruct.from_dict(item)
                children_container.append(item_tmp)
            
        
        # children
        children = None
        if item := d.get("children"):
            children = []
            for item in item:
                item_tmp = SimpleStruct.from_dict(item)
                children.append(item_tmp)
            
        return SimpleStruct(
            bool_value = bool_value,
            i8_value = i8_value,
            i64_value = i64_value,
            string_value = string_value,
            bytes_value = bytes_value,
            string_to_string = string_to_string,
            key_values = key_values,
            children_container = children_container,
            children = children,
        )
        
    

# KeyValue
KeyValue = typing.Type[typing.Dict[str, bytes]]

# Container
Container = typing.Type[typing.List["SimpleStruct"]]

# RealNumber
@dataclass
class RealNumber:
    real: typing.Optional[float] = None
    imagine: typing.Optional[float] = None

    def to_dict(self):
        result = {}
    
        # real
        result["real"] = self.real
    
        # imagine
        result["imagine"] = self.imagine
        return result
    

    @staticmethod
    def from_dict(d):
        
        # real
        real = d.get("real", None)
        
        # imagine
        imagine = d.get("imagine", None)
        return RealNumber(
            real = real,
            imagine = imagine,
        )
        
    

# Number
class Number(abc.ABC):
    pass

    @abc.abstractmethod
    def to_dict(self):
        pass
    
    @staticmethod
    def from_dict(d):
        type_ = d["type"]
        if type_ == "I64":
            payload = d["payload"]
            payload_tmp = payload
            return Number_I64(payload=payload_tmp)
        elif type_ == "F64":
            payload = d["payload"]
            payload_tmp = payload
            return Number_F64(payload=payload_tmp)
        elif type_ == "RealNumber":
            payload = d["payload"]
            payload_tmp = RealNumber.from_dict(payload)
            return Number_RealNumber(payload=payload_tmp)
        else:
            raise ValueError(f"invalid type: {type_}")
    

# variant I64 for Number
@dataclass
class Number_I64(Number):
    payload: int

    def to_dict(self):
        type_ = "I64"
        payload_tmp = self.payload
        return {
            "type": type_,
            "payload": payload_tmp,
        }


# variant F64 for Number
@dataclass
class Number_F64(Number):
    payload: float

    def to_dict(self):
        type_ = "F64"
        payload_tmp = self.payload
        return {
            "type": type_,
            "payload": payload_tmp,
        }


# variant RealNumber for Number
@dataclass
class Number_RealNumber(Number):
    payload: "RealNumber"

    def to_dict(self):
        type_ = "RealNumber"
        payload_tmp = self.payload.to_dict()
        return {
            "type": type_,
            "payload": payload_tmp,
        }


# BaseRequest
class BaseRequest(abc.ABC):
    pass

    @staticmethod
    @abc.abstractmethod
    def from_dict(d): pass

    @abc.abstractmethod
    def to_dict(self): pass

# AddRequest
@dataclass
class AddRequest(BaseRequest):
    request_id: typing.Optional[str] = None
    numbers: typing.Optional[typing.List["Number"]] = None

    def to_dict(self):
        result = {}
    
        # request_id
        result["request_id"] = self.request_id
    
        # numbers
        if self.numbers is None:
            result["numbers"] = None
        else:
            numbers_tmp = []
            for item in self.numbers:
                item_tmp = item.to_dict()
                numbers_tmp.append(item_tmp)
            
            result["numbers"] = numbers_tmp
        return result
    

    @staticmethod
    def from_dict(d):
        
        # request_id
        request_id = d.get("request_id", None)
        
        # numbers
        numbers = None
        if item := d.get("numbers"):
            numbers = []
            for item in item:
                item_tmp = Number.from_dict(item)
                numbers.append(item_tmp)
            
        return AddRequest(
            request_id = request_id,
            numbers = numbers,
        )
        
    

# ResetRequest
@dataclass
class ResetRequest(BaseRequest):
    request_id: typing.Optional[str] = None

    def to_dict(self):
        result = {}
    
        # request_id
        result["request_id"] = self.request_id
        return result
    

    @staticmethod
    def from_dict(d):
        
        # request_id
        request_id = d.get("request_id", None)
        return ResetRequest(
            request_id = request_id,
        )
        
    

File: examples/swift_package/Sources/SpecModel/example_spec.swift
import Foundation

public enum ModelError: Error {
    case Error
}

public struct SimpleStruct: Codable {
    public var bool_value: Bool
    public var i8_value: Int8
    public var i64_value: Int64?
    public var string_value: String?
    public var bytes_value: Data?
    public var string_to_string: [String:String]?
    public var key_values: SpecModel.KeyValue?
    public var children_container: SpecModel.Container?
    public var children: [SpecModel.SimpleStruct]?

    public init(bool_value: Bool, i8_value: Int8, i64_value: Int64? = nil, string_value: String? = nil, bytes_value: Data? = nil, string_to_string: [String:String]? = nil, key_values: SpecModel.KeyValue? = nil, children_container: SpecModel.Container? = nil, children: [SpecModel.SimpleStruct]? = nil) {
        self.bool_value = bool_value
        self.i8_value = i8_value
        self.i64_value = i64_value
        self.string_value = string_value
        self.bytes_value = bytes_value
        self.string_to_string = string_to_string
        self.key_values = key_values
        self.children_container = children_container
        self.children = children
    }
}

public typealias KeyValue = [String:Data]

public typealias Container = [SpecModel.SimpleStruct]

public struct RealNumber: Codable {
    public var real: Float64?
    public var imagine: Float64?

    public init(real: Float64? = nil, imagine: Float64? = nil) {
        self.real = real
        self.imagine = imagine
    }
}

public enum Number: Codable {
    case I64(Int64)
    case F64(Float64)
    case RealNumber(SpecModel.RealNumber)

    // coding keys
    enum CodingKeys: String, CodingKey {
        case type, payload
    }

    // decoder
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: CodingKeys.type)
        switch type {
            case "I64":
                let payload = try container.decode(Int64.self, forKey:.payload)
                self = .I64(payload)
            case "F64":
                let payload = try container.decode(Float64.self, forKey:.payload)
                self = .F64(payload)
            case "RealNumber":
                let payload = try container.decode(SpecModel.RealNumber.self, forKey:.payload)
                self = .RealNumber(payload)
            default:
                throw ModelError.Error
        }
    }

    // encoder
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
            case let .I64(payload):
                try container.encode("I64", forKey: .type)
                try container.encode(payload, forKey: .payload)
            case let .F64(payload):
                try container.encode("F64", forKey: .type)
                try container.encode(payload, forKey: .payload)
            case let .RealNumber(payload):
                try container.encode("RealNumber", forKey: .type)
                try container.encode(payload, forKey: .payload)
        }
    }
}

public protocol BaseRequest {
    var request_id: String? {
        get
        set
    }
}

public struct AddRequest: Codable, BaseRequest {
    public var request_id: String?
    public var numbers: [SpecModel.Number]?

    public init(request_id: String? = nil, numbers: [SpecModel.Number]? = nil) {
        self.request_id = request_id
        self.numbers = numbers
    }
}

public struct ResetRequest: Codable, BaseRequest {
    public var request_id: String?

    public init(request_id: String? = nil) {
        self.request_id = request_id
    }
}

public enum ConstInteger: Int64 {
    case Value1 = 1
    case Value2 = 2
}
